primary_beam_correction.py

```diff
@@ -194,27 +194,26 @@ def write_new_fits(pbc_image, path, outputFilename):
     images = read_fits(path)
     hdr = images.header
     newhdr = hdr.copy()
-    # add in required beam keywords
+    # change the frequency plane keywords, we don't want multiple frequency #axes
+    newhdr['CTYPE3'] = 'FREQ'
+    newhdr['NAXIS3'] = 1
+    newhdr['CDELT3'] = 1.0
     try:
-        if newhdr['CLEANBMJ'] < 0:
-            # add in required beam keywords
-            newhdr['BMAJ'] = float(newhdr['HISTORY'][-2][20:30])
-            newhdr['BMIN'] = float(newhdr['HISTORY'][-2][36:48])
-            newhdr['BPA'] = float(newhdr['HISTORY'][-2][55:])
-
-            newhdr['CTYPE3'] = 'FREQ'
-            newhdr['NAXIS3'] = 1
-            newhdr['CDELT3'] = 1.0
-            newhdr['CRVAL3'] = 1.284e9
-        else:
-            # add in required beam keywords
+        if 'CLEANBMJ' in newhdr:
+            if newhdr['CLEANBMJ'] < 0:
+                # Checking CLEANBMAJ in the history 
+                newhdr['BMAJ'] = float(newhdr['HISTORY'][-2][20:30])
+                newhdr['BMIN'] = float(newhdr['HISTORY'][-2][36:48])
+                newhdr['BPA'] = float(newhdr['HISTORY'][-2][55:])
+            else:
+                # add in required beam keywords
+                newhdr['BMAJ'] = newhdr['CLEANBMJ']
+                newhdr['BMIN'] = newhdr['CLEANBMN']
+                newhdr['BPA'] = newhdr['CLEANBPA']
+        elif 'BMAJ' in newhdr:
             newhdr['BMAJ'] = newhdr['CLEANBMJ']
             newhdr['BMIN'] = newhdr['CLEANBMN']
             newhdr['BPA'] = newhdr['CLEANBPA']
```

Comment:
This is now a bit of a mess.

Won't this work? And has the added advantage that you don't have to assume the locations of the cards in the HISTORY:
```
def _get_value_from_history(card, header):
    # Return the value of a card from the FITS HISTORY in header
    # Assumes card is found in a line of the HISTORY with format: 'card= value'
    for history in header['HISTORY']:
        line = history.replace('=', ' ').split()
        try:
            ind = line.index(card)
        except ValueError:
            continue
        return line[ind + 1]
    raise KeyError(f'{card} not found in HISTORY') 

try:
    if newhdr['CLEANBMJ'] > 0:
        # add in required beam keywords
        newhdr['BMAJ'] = newhdr['CLEANBMJ']
        newhdr['BMIN'] = newhdr['CLEANBMN']
        newhdr['BPA'] = newhdr['CLEANBPA']
    else:
        # Checking CLEANBMAJ in the history 
        newhdr['BMAJ'] = float(_get_value_from_history('BMAJ', newhdr))
        newhdr['BMIN'] = float(_get_value_from_history('BMIN', newhdr))
        newhdr['BPA'] = float(_get_value_from_history('BPA', newhdr))
except KeyError:
    logging.error('Exception occurred, keywords not found', exc_info=True)

Reply:
I see what you meant, initially, I did not fully get what you meant. I have incorporated your suggestions including your piece of code. It is working very well and I really like the extra feature of not assuming the index of the keywords in the header. Thanks.